### ISSUE TITLE
Why can't I use `model()[x]`to get value in some cases？

### ISSUE THREAD

**Issue by YsielX** [Z3Prover/z3#6450]
I'm using z3 to solve a simple constraint problem, but at the end I can't get the value through `model()[x]`
Error：
```
Traceback (most recent call last):
  File "d:\REVERSE\exam\newstar\ang.py", line 38, in <module>
    print(k[x])
  File "E:\python310\Lib\site-packages\z3\z3.py", line 6621, in __getitem__
    _z3_assert(False, "Integer, Z3 declaration, or Z3 constant expected")
  File "E:\python310\Lib\site-packages\z3\z3.py", line 107, in _z3_assert
    raise Z3Exception(msg)
z3.z3types.Z3Exception: Integer, Z3 declaration, or Z3 constant expected
```
The code looks like this
```python
from z3 import *
x = BitVec('x', 8)
x = ((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((x - 62) ^ 0xEB) - 51) ^ 0x71) \
                                                                             - 37) ^ 0xAA) \
                                                                           - 125) ^ 0x11) \
                                                                         + 53) ^ 0x6B) \
                                                                       - 123) ^ 0x4D) \
                                                                     - 70) ^ 0xF1) \
                                                                   - 102) ^ 0x56) \
                                                                 - 54) ^ 0x72) \
                                                               + 61) ^ 0x4A) \
                                                             - 4) ^ 0xDF) \
                                                           + 60) ^ 0x82) \
                                                         - 119) ^ 0xB1) \
                                                       - 65) ^ 0x4A) \
                                                     + 85) ^ 0x70) \
                                                   + 35) ^ 0xA4) \
                                                 - 74) ^ 0x61) \
                                               - 66) ^ 0xDF) \
                                             - 80) ^ 0x8E) \
                                           - 120) ^ 0x15) \
                                         + 51) ^ 0xC3) \
                                       + 74) ^ 0xA) \
                                     + 33) ^ 0x22) \
                                   - 100) ^ 0x9B) \
                                 - 57) ^ 9) \
                               + 56) ^ 0xA0) \
                             - 77) ^ 0x49) \
                           - 122) ^ 0xBB) \
                         - 102) ^ 0x7A) \
                       + 85) ^ 0x3B) \
+ 41 
s = Solver()
s.add(x==226)
s.check()
k=s.model()
print(k[x])
```
Could you tell me where is the problem?

**Comment by LeventErkok** [comment]
The problem here is the assignment to `x` on line 3, where you *redefine* `x`:

```python
x = ... x - 62 ...
```
 
So, what you did here is you took the "variable" `x` and turned it into an "expression". And you cannot index the model with an expression, only with variables.

The best way to solve this would be to *not* to redefine your variables like this: Declare them first, and assert constraints about them, use them in expressions, but do _not_ assign to them.

If you must assign to your variables, then you can use `evaluate` instead. Try:

```python
print(k.evaluate(x))
```

instead of

```python
print(k[x])
```

and it'll do the right thing.

TLDR: Don't redefine your symbolic variables to be expressions if you want to use them as an index to your model. This is a bad idea in any case, as it can cause confusion later on since you're losing access to the original inputs to your problem anyhow.

Side note: The stack-overflow forum for z3/z3py is more appropriate for these sorts of questions. See https://stackoverflow.com/questions/tagged/z3py

**Comment by YsielX** [comment]
Well, it works. Thanks!
